src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs

```diff
@@ -405,32 +405,43 @@ private void EmitLoweredConditionalAccessExpression(BoundLoweredConditionalAcces
-
-                    // if ((object)default(T) != null) 
-                    // {
-                    //     goto whenNotNull
-                    // }
-                    // else
-                    // {
-                    //     temp = receiverRef
-                    //     receiverRef = ref temp
-                    // }
-                    EmitDefaultValue(receiverType, true, receiver.Syntax);
-                    EmitBox(receiverType, receiver.Syntax);
-                    _builder.EmitBranch(ILOpCode.Brtrue, whenNotNullLabel);
-                    EmitLoadIndirect(receiverType, receiver.Syntax);
-
-                    cloneTemp = AllocateTemp(receiverType, receiver.Syntax);
-                    _builder.EmitLocalStore(cloneTemp);
-                    _builder.EmitLocalAddress(cloneTemp);
-                    _builder.EmitLocalLoad(cloneTemp);
-                    EmitBox(receiver.Type, receiver.Syntax);
-
-                    // here we have loaded a ref to a temp and its boxed value { &T, O }
+                    if (receiverTemp is null)
+                    {
+                        // unconstrained case needs to handle case where T is actually a struct.
+                        // such values are never nulls
+                        // we will emit a check for such case, but the check is really a JIT-time 
+                        // constant since JIT will know if T is a struct or not.
+
+                        // if ((object)default(T) != null) 
+                        // {
+                        //     goto whenNotNull
+                        // }
+                        // else
+                        // {
+                        //     temp = receiverRef
+                        //     receiverRef = ref temp
+                        // }
+                        EmitDefaultValue(receiverType, true, receiver.Syntax);
+                        EmitBox(receiverType, receiver.Syntax);
+                        _builder.EmitBranch(ILOpCode.Brtrue, whenNotNullLabel);
+                        EmitLoadIndirect(receiverType, receiver.Syntax);
+
+                        cloneTemp = AllocateTemp(receiverType, receiver.Syntax);
+                        _builder.EmitLocalStore(cloneTemp);
+                        _builder.EmitLocalAddress(cloneTemp);
+                        _builder.EmitLocalLoad(cloneTemp);
+                        EmitBox(receiverType, receiver.Syntax);
+
+                        // here we have loaded a ref to a temp and its boxed value { &T, O }
+                    }
+                    else
+                    {
+                        // we are calling the expression on a copy of the target anyway, 
+                        // so even if T is a struct, we don't need to make sure we call the expression on the original target.
+
+                        _builder.EmitLocalLoad(receiverTemp);
+                        EmitBox(receiverType, receiver.Syntax);
```

Comment:
This seems wrong. Why are we boxing unconditionally?

Reply:
We're boxing in order to do a null check. This is necessary since we could be dealing with a Nullable<T> or a reference type T.

Note that the IL for a ConditionalAccess on an unconstrained local also boxes:

```csharp
  .locals init (T V_0) //temp
  IL_0000:  ldarg.0
  IL_0001:  ldfld      ""T Foo<T>.t""
  IL_0006:  stloc.0
  IL_0007:  ldloc.0
  IL_0008:  box        ""T""
  IL_000d:  brtrue.s   IL_0012
  IL_000f:  ldnull
  IL_0010:  br.s       IL_001f
  IL_0012:  ldloca.s   V_0
  IL_0014:  constrained. ""T""
  IL_001a:  callvirt   ""string object.ToString()""
  IL_001f:  call       ""void System.Console.WriteLine(string)""
  IL_0024:  ret
```

Reply:
This is a little tricky because this code wasn't just decided on the IL, it was decided on the JITed assembly.

The above comment mentions "JIT time constant." What it's referring to is specifically the line `(object)default(T) == null`. The JIT specializes generic invocations, so at JIT time it's entirely decided what the actual type of `T` is. Thus, the JIT knows this expression is one of two things:

1. True, if `T` is a class or `Nullable<T>`
2. False if `T` is a standard value type

That means that *only one* of the if/else blocks will actually get compiled at JIT time. It *also* means that even though you see a `box` instruction in the IL, that `box` should never actually appear in the emitted machine code.

In your change, you're completely bypassing the condition and directly boxing the argument. I think this will functionally bypass the optimizations we're included to specifically avoided unnecessary boxing.

Reply:
@agocke 

I'm not sure I understand:

This is the IL we currently generate for a conditional access on a local:

```csharp

Reply:
I'm not sure what code you're examining. For the example

```C#
using System;
public class Foo<T> 
{
    readonly T t;
    
    public void Print()
    {
        Console.WriteLine(t?.ToString());
    }
}
```

it looks like the current IL is

```C#
.method public hidebysig 
        instance void Print () cil managed 
    {
        // Method begins at RVA 0x2050
        // Code size 62 (0x3e)
        .maxstack 2
        .locals init (
            [0] !T,
            [1] !T
        )

        IL_0000: ldarg.0
        IL_0001: ldfld !0 class Foo`1<!T>::t
        IL_0006: stloc.0
        IL_0007: ldloca.s 0
        IL_0009: ldloca.s 1
        IL_000b: initobj !T
        IL_0011: ldloc.1
        IL_0012: box !T
        IL_0017: brtrue.s IL_002d

        IL_0019: ldobj !T
        IL_001e: stloc.1
        IL_001f: ldloca.s 1
        IL_0021: ldloc.1
        IL_0022: box !T
        IL_0027: brtrue.s IL_002d

        IL_0029: pop
        IL_002a: ldnull
        IL_002b: br.s IL_0038

        IL_002d: constrained. !T
        IL_0033: callvirt instance string [mscorlib]System.Object::ToString()

        IL_0038: call void [mscorlib]System.Console::WriteLine(string)
        IL_003d: ret
    }
```

Are you examining different code?

Reply:
I'm looking at the IL for:

```csharp
public class Foo<T>
{
    private T t;
    
    public void Print() 
    {
        var temp = t;
        Console.WriteLine(temp?.ToString());
    }
}
```

Reply:
Ah, this makes sense. I missed that the ldloca.s just moved to the top, rather than disappeared. The key is that the value that's boxed is not the one actually used for the call, which it looks like this preserves.

I'll try to review the new changes today.